### ISSUE TITLE
Cluster creation must be reliable

### ISSUE THREAD

**Issue by alex-mohr** [GoogleCloudPlatform/kubernetes#4064]
Cluster creation -- at least for GCE as a provider -- has a 1-2% failure rate observed by jenkins e2e tests.  We should make it more reliable.

One approach would be to open source the GKE cluster creation code.

**Comment by satnam6502** [comment]
It used to be **much** worse :-)
We've added a lot of code to harden against gcloud errors ('Baclend Error'); code to try to re-run the Docker service; code to more tolerantly check the Kubelet status. We sometimes still timeout a bit too early (e.g. checking Kubelet health) so perhaps we should just increase a timeout and that would maybe buy as 1%. Over the 101 runs I observed there were a couple of situations where the cluster bring up process just died abruptly without any useful error message. We should add more diagnostics and work out what actually went wrong so we can fix it.
